### PR TITLE
feat: Support Context in Assertions API

### PR DESCRIPTION
**Pull request by smeadows-okta**
The assertions API should allow specifying check requests with context.

## Description

Added support assertion API for contextual tuples and context.
```
{
    "assertions": [
        {
            "tuple_key": {
                "object": "folder:test-folder",
                "relation": "viewer",
                "user": "user:anne"
            },
            "expectation": true,
            "contextual_tuples": {
                "tuple_keys": [
                    {
                        "user": "user:anne",
                        "relation": "viewer",
                        "object": "folder:test-folder"
                    }
                ]
            }
        }
    ]
}

curl -v -X PUT --header "Accept: */*" --header "Content-Type: application/json" -d @./no_condition.json http://localhost:8080/stores/01HYE30RYJ6ZJ3ZCD1ZWHJD2KJ/assertions/01HYE316Z4CNBFKFC8E9T45SXV
* Host localhost:8080 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8080...
* Connected to localhost (::1) port 8080
> PUT /stores/01HYE30RYJ6ZJ3ZCD1ZWHJD2KJ/assertions/01HYE316Z4CNBFKFC8E9T45SXV HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.6.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 515
> 

With Condition json:
{
    "assertions": [
        {
            "tuple_key": {
                "object": "folder:test-folder",
                "relation": "viewer",
                "user": "user:anne"
            },
            "expectation": true,
            "contextual_tuples": {
                "tuple_keys": [
                    {
                        "user": "user:anne",
                        "relation": "viewer",
                        "object": "folder:test-folder",
                        "condition": {
                            "name": "name_starts_with_alpha",
                            "context": {
                                "folder_name": "test"
                            }
                        }
                    }
                ]
            }
        }
    ]
}

curl steven.meadows$ curl -v -X PUT --header "Accept: */*" --header "Content-Type: application/json" -d @./condition.json http://localhost:8080/stores/01HYE3SW24P34DGXZ7ZF90GP0G/assertions/01HYE3T88BFZXBQZSHHTW4Z38W
* Host localhost:8080 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8080...
* Connected to localhost (::1) port 8080
> PUT /stores/01HYE3SW24P34DGXZ7ZF90GP0G/assertions/01HYE3T88BFZXBQZSHHTW4Z38W HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.6.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 762
> 
< HTTP/1.1 204 No Content
< Content-Type: application/json
< Openfga-Authorization-Model-Id: 01HYE3T88BFZXBQZSHHTW4Z38W
< Openfga-Store-Id: 01HYE3SW24P34DGXZ7ZF90GP0G
< Vary: Origin
< X-Http-Code: 204
< X-Request-Id: 4ca8118a-5807-4382-90b8-f0486175a2e6
< Date: Tue, 21 May 2024 17:45:15 GMT
< 
* Connection #0 to host localhost left intact
```
## References
https://github.com/openfga/openfga/issues/1353
https://github.com/openfga/api/pull/157

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/server/test/write_assertions.go [hunk 1]
```diff
@@ -22,41 +22,54 @@ import (
...
+				define reader: [user]
+				define can_read: reader`,
 			assertions: []*openfgav1.Assertion{{
 				TupleKey:    tuple.NewAssertionTupleKey("repo:test", "reader", "user:elbuo"),
 				Expectation: false,
 			}},
 		},
 		{
 			_name: "writing_assertions_succeeds_when_it_is_not_directly_assignable",
+			//	model: "writing_assertions_succeeds",
```
Comment (jpadilla) [inline pkg/server/test/write_assertions.go:49]:
```suggestion
```

File: internal/validation/validation.go [hunk 2]
```diff
@@ -31,6 +31,14 @@ func ValidateUserObjectRelation(typesys *typesystem.TypeSystem, tk *openfgav1.Tu
 	return nil
 }
 
+func ValidateUserObjectRelationCondition(typesys *typesystem.TypeSystem, tk *openfgav1.TupleKey) error {
```
Comment (jpadilla) [inline internal/validation/validation.go:34]: can we add a comment to this new function?

File: pkg/server/commands/write_assertions.go [hunk 3]
```diff
@@ -66,8 +66,13 @@ func (w *WriteAssertionsCommand) Execute(ctx context.Context, req *openfgav1.Wri
...
 		for _, ct := range assertion.GetContextualTuples().GetTupleKeys() {
-			if err = validation.ValidateUserObjectRelation(typesys,
-				tupleUtils.NewTupleKey(ct.GetObject(), ct.GetRelation(), ct.GetUser())); err != nil {
+			if err = validation.ValidateUserObjectRelationCondition(typesys,
+				tupleUtils.NewTupleKeyWithCondition(
+					ct.GetObject(),
+					ct.GetRelation(),
+					ct.GetUser(),
+					ct.GetCondition().GetName(),
+					ct.GetCondition().GetContext())); err != nil {
```
Comment (jpadilla) [inline pkg/server/commands/write_assertions.go:75]: is there the possibility of `ct.GetCondition()` returning `nil` here and causing a panic?

**Review by jpadilla — CHANGES_REQUESTED** [review]

File: pkg/server/commands/write_assertions.go (hunk 3 above)
Comment (smeadows-okta) [reply pkg/server/commands/write_assertions.go:75]: No, gRPC checks and return nil for both "GetCondition" and "GetName or GetContext"

File: internal/validation/validation.go (hunk 2 above)
Comment (smeadows-okta) [reply internal/validation/validation.go:36]: 👍

File: pkg/server/test/write_assertions.go (hunk 1 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go]: 👍

**Review by jpadilla — DISMISSED** [review]

File: pkg/server/commands/write_assertions.go (hunk 3 above)
Comment (miparnisari) [inline pkg/server/commands/write_assertions.go:75]: why not `validation.ValidateUserObjectRelationCondition(typesys, ct)`? it passes the tests

**Comment by jon-whit** [comment]
Regarding the changes to the API in https://github.com/openfga/api/pull/157. This line https://github.com/openfga/api/pull/157/files#diff-d3a74ea9ecaac0a46611253dc0c4bc9c402e2ad9eb255cd41d07266453e258f0R1564 we should avoid duplicating that stuttered field definition. Since this is a new addition to the API, let's follow the same pattern we did in ListUsers. See https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L877. It stutters less. There's no need for the extra level of indirection.

File: pkg/server/commands/write_assertions.go (hunk 3 above)
Comment (smeadows-okta) [reply pkg/server/commands/write_assertions.go:75]: Corrected!

**Comment by smeadows-okta** [comment]
> Regarding the changes to the API in [openfga/api#157](https://github.com/openfga/api/pull/157). This line https://github.com/openfga/api/pull/157/files#diff-d3a74ea9ecaac0a46611253dc0c4bc9c402e2ad9eb255cd41d07266453e258f0R1564 we should avoid duplicating that stuttered field definition. Since this is a new addition to the API, let's follow the same pattern we did in ListUsers. See https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L877. It stutters less. There's no need for the extra level of indirection.

Since it's been merged already, I will need to open another PR to fix.

**Comment by jon-whit** [comment]
This PR is titled "Support Context in Assertions API" but we're not adding Context to to the Assertions API. There is no corresponding PR in openfga/api that adds `context` field to the [Assertion](https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L1552). We need to add https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L823 to the `Assertion` definition.

**Pull request closed without merging**